<a href="https://colab.research.google.com/github/NicoRearte03/Data_Science_I/blob/main/Pre_Entrega_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import os

# Configurar el token directamente en el entorno de Colab
os.environ['KAGGLE_API_TOKEN'] = "KGAT_98eb3f4053f88dd76a93450be80e49bb"

print("¡Token de Kaggle configurado!")


¡Token de Kaggle configurado!


In [ ]:
# 1. Descargar el dataset usando la API de Kaggle
!KAGGLE_API_TOKEN=KGAT_98eb3f4053f88dd76a93450be80e49bb kaggle datasets download -d sergionefedov/fraud-detection-1m-transactions-7-fraud-types

# 2. Descomprimir los archivos
import zipfile

with zipfile.ZipFile('fraud-detection-1m-transactions-7-fraud-types.zip', 'r') as zip_ref:
    zip_ref.extractall('dataset')

print("¡Listo! Dataset descargado y descomprimido en la carpeta /dataset")

Dataset URL: https://www.kaggle.com/datasets/sergionefedov/fraud-detection-1m-transactions-7-fraud-types
License(s): apache-2.0
fraud-detection-1m-transactions-7-fraud-types.zip: Skipping, found more recently modified local copy (use --force to force download)
¡Listo! Dataset descargado y descomprimido en la carpeta /dataset


In [ ]:
# =========================================================
# BLOQUE 1: INGESTA DE DATOS Y PRIMER VISTAZO
# =========================================================
import pandas as pd

# Carga del dataset principal de transacciones
df = pd.read_csv("dataset/transactions.csv")

# Muestra estilizada de las primeras 5 filas
display(df.head(5))

,transaction_id,account_id,timestamp,hour_of_day,day_of_week,is_weekend,amount,merchant_category,mcc_code,merchant_country,...,ip_risk_score,is_foreign_txn,time_since_last_s,velocity_1h,amount_vs_avg_ratio,account_age_days,has_2fa,credit_limit,is_fraud,fraud_pattern
0,TXN000000001,ACC0016173,2023-02-21 08:02:38,8,1,0,168.42,travel,4511,CA,...,53.2,1,21,3,2.6423,3256,1,3958.46,0,NaN
1,TXN000000002,ACC0011196,2024-05-12 23:13:34,23,6,1,85.78,online_retail,5999,AU,...,25.3,1,234,1,0.7279,1527,1,3553.35,0,NaN
2,TXN000000003,ACC0001181,2023-09-22 23:28:21,23,4,0,20.15,pharmacy,5912,CA,...,21.3,1,85,1,0.1851,2230,1,4362.57,0,NaN
3,TXN000000004,ACC0037105,2022-09-28 23:26:38,23,2,0,62.49,grocery,5411,US,...,13.7,0,98,0,1.5223,1863,1,3194.84,0,NaN
4,TXN000000005,ACC0028471,2023-02-23 17:54:13,17,3,0,71.68,online_retail,5999,US,...,9.7,0,721,2,0.7724,1728,0,11850.06,0,NaN


In [ ]:
# =========================================================
# BLOQUE 2: RADIOGRAFÍA TÉCNICA DEL DATASET
# =========================================================

# 1. Dimensiones exactas (filas y columnas)
print(f"Dimensiones del dataset: {df.shape}")
print(f"Total de filas: {df.shape[0]:,}")
print(f"Total de columnas: {df.shape[1]}")
print("-" * 50)

# 2. Tipos de datos por columna e información estructural
df.info()

# 3. Resumen estadístico de variables numéricas
display(df.describe())

Dimensiones del dataset: (1000000, 23)
Total de filas: 1,000,000
Total de columnas: 23
--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 23 columns):
 #   Column               Non-Null Count    Dtype  
---  ------               --------------    -----  
 0   transaction_id       1000000 non-null  object 
 1   account_id           1000000 non-null  object 
 2   timestamp            1000000 non-null  object 
 3   hour_of_day          1000000 non-null  int64  
 4   day_of_week          1000000 non-null  int64  
 5   is_weekend           1000000 non-null  int64  
 6   amount               1000000 non-null  float64
 7   merchant_category    1000000 non-null  object 
 8   mcc_code             1000000 non-null  int64  
 9   merchant_country     1000000 non-null  object 
 10  card_present         1000000 non-null  int64  
 11  device_type          1000000 non-null  object 
 12  device_known     

,hour_of_day,day_of_week,is_weekend,amount,mcc_code,card_present,device_known,ip_risk_score,is_foreign_txn,time_since_last_s,velocity_1h,amount_vs_avg_ratio,account_age_days,has_2fa,credit_limit,is_fraud
count,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000
mean,11.494197,3.003311,0.287152,183.738264,5703.713286,0.350654,0.892337,21.605828,0.294417,180.112610,1.050421,3.298512,1833.588908,0.647567,6706.938185,0.017143
std,6.920962,2.002666,0.452433,316.693903,548.917957,0.477175,0.309954,16.269652,0.455780,180.148636,1.095055,7.157055,1046.489932,0.477728,6174.164106,0.129804
min,0.000000,0.000000,0.000000,1.830000,4511.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000,0.011800,30.000000,0.000000,500.000000,0.000000
25%,5.000000,1.000000,0.000000,43.480000,5411.000000,0.000000,1.000000,8.500000,0.000000,52.000000,0.000000,0.597100,926.000000,0.000000,2833.750000,0.000000
50%,11.000000,3.000000,0.000000,81.140000,5734.000000,0.000000,1.000000,18.700000,0.000000,125.000000,1.000000,1.285300,1832.000000,1.000000,4880.100000,0.000000
75%,17.000000,5.000000,1.000000,189.760000,5999.000000,1.000000,1.000000,30.700000,1.000000,250.000000,2.000000,3.155500,2738.000000,1.000000,8384.040000,0.000000
max,23.000000,6.000000,1.000000,25000.000000,7995.000000,1.000000,1.000000,100.000000,1.000000,2733.000000,15.000000,945.537100,3649.000000,1.000000,50000.000000,1.000000


### Hallazgos Estadísticos y Anomalías Detectadas:

1. **Sesgo y presencia de outliers en la distribución de montos:**  
   Al analizar los valores del `describe()`, se observa una marcada diferencia entre la mediana (percentil 50%) y el valor máximo en las variables de montos (`amount` / `amount_ratio`). Esto indica un sesgo positivo (hacia la derecha) y la presencia de *outliers* significativos, lo cual es característico de operaciones financieras atípicas o posibles eventos de fraude.

2. **Marcado desbalance en la variable objetivo (`is_fraud`):**  
   La media de la columna target se sitúa cercana a 0.017 (1.71%), lo que confirma que las transacciones fraudulentas constituyen una minoría absoluta en el conjunto de datos. Este desbalance refleja fielmente un entorno de producción real, pero requerirá métricas específicas de evaluación más allá del simple *accuracy*.

In [ ]:
# =========================================================
# BLOQUE 3: DIAGNÓSTICO DE DATOS FALTANTES
# =========================================================

# 1. Calcular el porcentaje exacto de nulos por columna
porcentaje_nulos = (df.isnull().sum() / len(df)) * 100

# 2. Ordenar de mayor a menor
porcentaje_nulos_ordenado = porcentaje_nulos.sort_values(ascending=False)

# 3. Crear un DataFrame empaquetado para mostrarlo prolijo en pantalla
df_diagnostico_nulos = pd.DataFrame({
    'Columna': porcentaje_nulos_ordenado.index,
    '% Datos Nulos': porcentaje_nulos_ordenado.values
})

# Mostrar la tabla formateada
display(df_diagnostico_nulos)

,Columna,% Datos Nulos
0,fraud_pattern,98.2857
1,account_id,0.0000
2,transaction_id,0.0000
3,hour_of_day,0.0000
4,day_of_week,0.0000
5,is_weekend,0.0000
6,timestamp,0.0000
7,amount,0.0000
8,merchant_category,0.0000
9,merchant_country,0.0000


In [ ]:
# =========================================================
# BLOQUE 4: LÓGICA ALGORÍTMICA Y FUNCIONES PERSONALIZADAS
# =========================================================

# 1. FUNCIÓN FORMAL (def)
# Evaluamos la columna 'amount_vs_avg_ratio' para categorizar el nivel de sospecha según el desvío del gasto
def clasificar_nivel_riesgo(ratio):
    """
    Clasifica el riesgo de una transacción comparando el monto contra
    el promedio histórico del usuario.
    """
    if ratio > 5.0:
        return 'Alto Riesgo'
    elif ratio > 2.0:
        return 'Riesgo Medio'
    else:
        return 'Riesgo Bajo'

# Aplicamos la función formal con .apply() para crear una nueva columna
df['nivel_riesgo_monto'] = df['amount_vs_avg_ratio'].apply(clasificar_nivel_riesgo)


# 2. FUNCIÓN EXPRESS (lambda)
# Convertimos el monto original ('amount') de unidades a miles para simplificar la lectura
df['amount_en_miles'] = df['amount'].apply(lambda x: round(x / 1000, 2))


# Muestra de verificación con las columnas originales y las dos nuevas generadas
display(df[['amount', 'amount_vs_avg_ratio', 'nivel_riesgo_monto', 'amount_en_miles']].head(10))

,amount,amount_vs_avg_ratio,nivel_riesgo_monto,amount_en_miles
0,168.42,2.6423,Riesgo Medio,0.17
1,85.78,0.7279,Riesgo Bajo,0.09
2,20.15,0.1851,Riesgo Bajo,0.02
3,62.49,1.5223,Riesgo Bajo,0.06
4,71.68,0.7724,Riesgo Bajo,0.07
5,120.73,1.8660,Riesgo Bajo,0.12
6,150.13,3.2356,Riesgo Medio,0.15
7,59.84,0.5575,Riesgo Bajo,0.06
8,47.24,0.2998,Riesgo Bajo,0.05
9,49.31,2.2373,Riesgo Medio,0.05
